### PR TITLE
MTDSA-17724 Parser-validator redesign

### PR DESCRIPTION
**Pull request by mattstephens-hmrc**

### CODE REVIEW

**Comment by jeremystone** [comment]
Looks good. Just a few minor things that might be worth considering...

- It isn’t obvious (at least to me, without looking at the implementation) what the `RulesValidator` `combineResults` and `combineProgress` methods do. And for `combineResults` it takes many parameters and it isn’t immediately clear that the first one is special. Could the first parameter be removed so that there is only one flavour (i.e. `combineProgress` but maybe it can be renamed: `combine`, or `validateAll`). So rather than:

```scala
combineResults(parsed, rule1, rule2, ...)
```
you could have:
```scala
validateAll(rule1, rule2, ...).map(_ => parsed)
```
or with an `onSuccess` (or `onValid` or  `ifValid` or something else?)  extension method in the `RulesValidator` trait that does the `map(_ => parsed)`:
```scala
validateAll(rule1, rule2, ...).onSuccess(parsed)
```

- The only other thing that might confuse is the `RulesValidator` itself. The name would imply it is a kind of `Validator` (and it is defined in the same file), but it isn’t. It might just be a case of renaming, but, playing Devil’s advocate here, is there a need for a `RulesValidator` to be a type at all? The `RulesValidator` instances don’t appear to be separately unit tested so the ‘interface’ could just be implemented (ad hoc) as a private method within any `Validator` that uses it. A trait providing combineResults/combineProgress (or whatever — see above) could still be mixed in as required, or part of the `Validator` trait.

File: test/v3/controllers/validators/SubmitSelfEmploymentBsasValidatorFactorySpec.scala [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        )
+      }
+    }
+
+    "return ValueFormatError" when {
+      "single fields are invalid" when {
+        List(
+          "/income/turnover",
+          "/income/other",
+          // "/expenses/consolidatedExpenses",
```
Comment (SamuelJacobsHMRC) [inline test/v3/controllers/validators/SubmitSelfEmploymentBsasValidatorFactorySpec.scala:195]: can we delete the comment?

File: app/api/controllers/validators/resolvers/Resolver.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import cats.data.Validated.Valid
+
+/** Parses a raw value (e.g. String or JsValue) to a target type, validating in the process.
+  *
+  * @tparam S
+  *   The source data type
+  * @tparam T
+  *   The target data type
+  */
+trait Resolver[S, T] {
```
Comment (SamuelJacobsHMRC) [inline app/api/controllers/validators/resolvers/Resolver.scala:30]: Is it not worth having a test file for this?

File: app/api/controllers/RequestHandler.scala [hunk 3]
```diff
@@ -136,33 +133,34 @@ object RequestHandler {
...
+          apiVersion: Version): Future[Result] = {
 
         logger.info(
           message = s"[${ctx.endpointLogContext.controllerName}][${ctx.endpointLogContext.endpointName}] " +
             s"with correlationId : ${ctx.correlationId}")
 
         val result =
           for {
-            parsedRequest <- EitherT.fromEither[Future](parser.parseRequest(rawData))
+            parsedRequest   <- EitherT.fromEither[Future](validator.validateAndWrapResult())
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/api/controllers/RequestHandler.scala:159]: The method isn't a pure function, so should retain the ()

File: app/api/controllers/validators/resolvers/ResolveNino.scala [hunk 4]
```diff
@@ -14,18 +14,23 @@
...
   private val ninoRegex =
-    "^([ACEHJLMOPRSWXY][A-CEGHJ-NPR-TW-Z]|B[A-CEHJ-NPR-TW-Z]|G[ACEGHJ-NPR-TW-Z]|" +
-      "[KT][A-CEGHJ-MPR-TW-Z]|N[A-CEGHJL-NPR-SW-Z]|Z[A-CEGHJ-NPR-TW-Y])[0-9]{6}[A-D ]?$"
+    ("^([ACEHJLMOPRSWXY][A-CEGHJ-NPR-TW-Z]|B[A-CEHJ-NPR-TW-Z]|G[ACEGHJ-NPR-TW-Z]|" +
+      "[KT][A-CEGHJ-MPR-TW-Z]|N[A-CEGHJL-NPR-SW-Z]|Z[A-CEGHJ-NPR-TW-Y])[0-9]{6}[A-D ]?$").r
 
-  def validate(nino: String): List[MtdError] = {
-    if (Nino.isValid(nino) && nino.matches(ninoRegex)) NoValidationErrors else List(NinoFormatError)
+  def apply(value: String, error: Option[MtdError], path: Option[String]): Validated[Seq[MtdError], Nino] = {
+    if (Nino.isValid(value) && ninoRegex.matches(value))
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/api/controllers/validators/resolvers/ResolveNino.scala:31]: Where are the warnings?

File: app/api/controllers/validators/resolvers/ResolveParsedNumber.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import cats.data.Validated
+import cats.data.Validated.{ Invalid, Valid }
+
+case class ResolveParsedNumber(min: BigDecimal = 0, max: BigDecimal = 99999999999.99, disallowZero: Boolean = false)
+    extends Resolver[BigDecimal, BigDecimal] {
+
+  def apply(value: BigDecimal, notUsedError: Option[MtdError], path: Option[String]): Validated[Seq[MtdError], BigDecimal] = {
+    val usePath = requirePath(path)
+
+    if (value >= min && value <= max && value.scale <= 2 && (!disallowZero || value != 0))
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/api/controllers/validators/resolvers/ResolveParsedNumber.scala:29]: Normally I'd agree, but in this case it's really outside the scope of the change. Idea was to migrate the existing Validators and Validations, not to rewrite them all!  (Maybe one day....)

File: app/api/controllers/validators/resolvers/Resolver.scala (hunk 2 above)
Comment (mattstephens-hmrc) [reply app/api/controllers/validators/resolvers/Resolver.scala:30]: sure, will add one

File: app/api/controllers/validators/resolvers/ResolveNino.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (SamuelJacobsHMRC) [reply app/api/controllers/validators/resolvers/ResolveNino.scala:31]: I'm getting them in from IntelliJ. Always happens when we have a multiline if block

File: app/api/controllers/validators/resolvers/ResolveParsedNumber.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (SamuelJacobsHMRC) [reply app/api/controllers/validators/resolvers/ResolveParsedNumber.scala:29]: i find them a little distraction but not the end of the world.. your call!

File: app/api/controllers/validators/Validator.scala [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+trait RulesValidator[PARSED] extends Logging {
+
+  protected val valid: Validated[Seq[MtdError], Unit] = Valid(())
+
+  def validateBusinessRules(parsed: PARSED): Validated[Seq[MtdError], PARSED]
+
+  protected def combineResults(parsed: PARSED, results: Validated[Seq[MtdError], _]*): Validated[Seq[MtdError], PARSED] =
+    results.traverse_(identity).map(_ => parsed)
+
+  protected def combineProgress(results: Validated[Seq[MtdError], _]*): Validated[Seq[MtdError], Unit] =
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/api/controllers/validators/Validator.scala:34]: They need to be separate, as they have different return types, and different intents (reflected in the names)

File: app/api/controllers/validators/Validator.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            Left(ErrorWrapper(correlationId, err, None))
+
+          case errs =>
+            logger.warn(s"Validation failed with ${errs.map(_.code).mkString(",")} error for the request with CorrelationId: $correlationId")
+            Left(ErrorWrapper(correlationId, BadRequestError, Some(errs)))
+        }
+    }
+  }
+
+  private def combineErrors(errors: Seq[MtdError]): Seq[MtdError] = {
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/api/controllers/validators/Validator.scala:63]: Again that isn't my code, the goal here isn't to rewrite the entire API (though I would very much like to one day!), but just to refactor the parser+validator design

File: app/v3/controllers/SubmitSelfEmploymentBsasController.scala [hunk 8]
```diff
@@ -33,16 +32,16 @@ import javax.inject.{Inject, Singleton}
...
+                                                    validatorFactory: SubmitSelfEmploymentBsasValidatorFactory,
+                                                    service: SubmitSelfEmploymentBsasService,
+                                                    nrsService: SubmitSelfEmploymentBsasNrsProxyService,
+                                                    hateoasFactory: HateoasFactory,
+                                                    auditService: AuditService,
+                                                    cc: ControllerComponents,
+                                                    val idGenerator: IdGenerator)(implicit ec: ExecutionContext, appConfig: AppConfig)
+    extends AuthorisedController(cc)
     with V3Controller
     with Logging {
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/v3/controllers/SubmitSelfEmploymentBsasController.scala:46]: outside the scope of the PR

File: app/v3/models/request/AccountingPeriod.scala [hunk 9]
```diff
@@ -16,7 +16,7 @@
 
 package v3.models.request
 
-import play.api.libs.json.{Json, OFormat}
+import play.api.libs.json.{ Json, OFormat }
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/v3/models/request/AccountingPeriod.scala:19]: Agreed, but really outside the scope of the PR

File: app/v3/services/ListBsasService.scala [hunk 10]
```diff
@@ -21,43 +21,45 @@ import api.models.errors._
...
-class ListBsasService @Inject()(connector: ListBsasConnector) extends BaseService {
+class ListBsasService @Inject() (connector: ListBsasConnector) extends BaseService {
+
+  def listBsas(
+      request: ListBsasRequestData
+  )(implicit ctx: RequestContext, ec: ExecutionContext): Future[ServiceOutcome[ListBsasResponse[BsasSummary]]] = {
+
+    connector
+      .listBsas(request)
+      .map(_.leftMap(mapDownstreamErrors(errorMap)))
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/v3/services/ListBsasService.scala:39]: Yeah I was in two minds about whether to amp up the cats or not. Personally I prefer seeing that there's nested mapping taking place, but it depends. In this particular case, notice that the "bimap" version doesn't reduce the amount of code at all, and is slightly less clear if anything.

File: test/api/controllers/RequestParserSpec.scala [hunk 11]
```diff
@@ -1,81 +0,0 @@
-/*
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply test/api/controllers/RequestParserSpec.scala:1]: yeah agreed, I think I deleted their tests before reintroducing RequestParser etc in v2.

File: app/api/controllers/validators/Validator.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeremystone) [reply app/api/controllers/validators/Validator.scala:34]: I guess I think the thing is that in the case of `combineResults`, other than it being the first parameter, `parsed` is not sufficiently distinguished from the list of validations/rules that follow it to make its role perfectly clear. An extension method with suitable naming (not sure what, though, but see earlier comment) would enable `parsed` to stand out. (And you would only need the one method, unless I'm missing something.)

**Review by SamuelJacobsHMRC — APPROVED** [review]

**Pull request merged**